### ISSUE TITLE
Problems when folder and/or home disk are on extremely large filesystem?

### ISSUE THREAD

**Issue by josefsachsconning** [syncthing/syncthing#3930]
(when `Minimum Free Disk Space` in folder advanced settings is set to default value of 1)
```
2017-01-24 13:34:16: Stopping folder "notes" (uxhud-2egzw) - folder has insufficient free space
```

(when Actions > Advanced > Options > minHomeDiskFreePct is set to default value of 1)
```
2017-01-24 13:42:53: Stopping folder "notes" (uxhud-2egzw) - home disk has insufficient free space
```

Works okay as long as I change both values to 0.
Here is filesystem information:
```
$ df .
Filesystem                                              1K-blocks     Used        Available Use% Mounted on
fs-whatever.efs.us-east-1.amazonaws.com://foobar 9007199254740992 44992512 9007199209748480   1% /home/foobar
```

### Version Information

Syncthing Version: v0.14.20, Linux (64 bit)
OS Version: Ubuntu 14.04

**Comment by calmh** [comment]
Large, and more than 99% full. Fractional values are also allowed, such as 0.1 if you want to go up to 99.9% full. I don't see an issue here except possibly with the default value?

**Comment by josefsachsconning** [comment]
It looks almost empty to me.

**Comment by calmh** [comment]
Yeah, my first comment was incorrect because I didn't read properly and I deleted it. Not quick enough I guess. :)

9223372036854775807 is the maximum int64 value (2^63). That filesystem is 9007199254740992*1024 = 9223372036854775808 bytes, or thereabout, with rounding. In other words it's possible it exceeds an int64.

**Comment by josefsachsconning** [comment]
I would have guessed that it would use an unsigned int.

**Comment by calmh** [comment]
It might, but we don't.

**Comment by AudriusButkevicius** [comment]
I think you can just set minHomeDiskFreePct  to 0?

**Comment by calmh** [comment]
Yes, even says so in the initial report. Lets not torment this guy any more by not reading his issue :)

**Comment by AudriusButkevicius** [comment]
This is what happens when a report contains more than a few lines :(

Sorry.

**Comment by josefsachsconning** [comment]
Not complaining, really appreciate the help, but let me know if you prefer less information in future reports (if any).

**Comment by calmh** [comment]
So the statvfs API gives us size of blocks and number of blocks. It looks like this filesystem synthesizes it's size to be exactly 2^63. That exceeds int64 when we do the multiplication, so we get a negative total size and things fall apart from there.

**Comment by calmh** [comment]
@josefsachsconning bug report is fine, you just managed to catch two potentially tired maintainers skimming lots of issues quickly. I know at least one of us resembles that description...

**Comment by Ferroin** [comment]
This is going to sound odd, but how important is high accuracy for this calculation?  If it's not that important, it might make more sense to do the usage calculation on the block counts instead of the byte counts.  This would also give a high enough overflow point that pretty much anything other than ZFS (which is insane and supports 2^128 byte filesystems...) and some distributed filesystems (which don't actually use blocks in the same way as most local filesystems) will hit filesystem limits before they hit the overflow point in this calculation.  The downside is that the values used by Syncthing will be inconsistent with how some utilities report space usage (although most such tools are arguably wrong because most filesystems don't do tail-packing, and therefore can't split a block among multiple files, which in turn means that block usage is a better indicator of free space than bytes used on most filesystems), but the discrepancy won't be more than a few hundredths of a percent for most people (the exact value is proportionate to the number of partially used filesystem blocks and how full each one is).

**Comment by calmh** [comment]
Yeah. For the most part we're just interested in the free space percentage. For that we don't care if the disk space unit is bytes, megabytes, blocks, whatever. There are a couple of places where we do check specifically for the amount of free space in bytes and compare it with a file we are going to sync, to see if it will fit. There we need some kind of known size.

But we could standardize on reporting and calculating in for example 128k blocks and not be too far off the truth. We would right shift the block count by 17 bits before doing the multiplication, buying 17 more bits until exhaustion.

**Comment by myphsto** [comment]
Could you not enhance du to return MB or GB instead of only KB?  This way when du populates from the OS call you can still use the int64 but understand the value returned is in MB or GB.  Passing the units also can catch the int64 overrun in du for very large values and return MB or GB instead.

**Comment by rdebath** [comment]
Sounds complicated. Why not just do the calculation using the floating point `double` type.
You don't need per byte accuracy for large values, but you do get per byte accuracy for values under a petabyte.
On a modern FPU the performance of a single `double` operation is just a little bit slower than an int64 and often faster than an int128.

**Comment by pchico83** [comment]
Hi, we are also facing this issue on Okteto (https://github.com/okteto/okteto) when synchronizing to large filesystems on ML scenarios
